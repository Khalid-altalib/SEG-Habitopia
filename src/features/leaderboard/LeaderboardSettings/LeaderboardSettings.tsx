import { Button, HStack, Select } from "native-base";
import React from "react";
import { Challenge } from "../../../../types";
import { useDispatch, useSelector } from "../../../app/hooks";
import { fetchChallenges } from "../../challenges/challengesSlice";
import {
  changeSetting,
  fetchLeaderboard,
  LeaderboardState,
} from "../leaderboardSlice";
import { useState, useEffect } from "react";
import StatusContainer from "@components/StatusContainer/StatusContainer";

type Props = {};

const timeIntervals = ["Weekly", "Monthly", "All Time"];

const LeaderboardSettings = (props: Props) => {
  const timeInterval = useSelector((state) => state.leaderboard.timeInterval);

  // const challenges = useSelector((state) => state.challenges.challenges); BACKEND_PLACEHOLDER

  // const challengeTypes = challenges.map(
  //   (challenge: Challenge) => challenge.name
  // );

  const { challenges, fetchChallenges: requestStatus } = useSelector(
    (state) => state.challenges
  );
  const { loading, error } = requestStatus;

  const [selectedChallengeType, setSelectedChallengeType] = useState("");

  const dispatch = useDispatch();

  const handleFetchChallenges = async () => {
    await dispatch(fetchChallenges());
    setSelectedChallengeType(challenges[0].name);
  };

  useEffect(() => {
    handleFetchChallenges();
  }, []);

  const handleChange = async (settingName: string, value: string) => {
    dispatch(changeSetting({ name: settingName, value: value }));
    await dispatch(fetchLeaderboard());
  };

  useEffect(() => {
    if (selectedChallengeType) {
      handleChange("challengeType", selectedChallengeType);
    }
  }, [selectedChallengeType]);

  return (
    <StatusContainer loading={loading} error={error} data={challenges}>
      <HStack space={4} testID="leaderboard-settings">
        {challenges &&
          selectedChallengeType !== "" &&
          challenges.map((challenge) => {
            const challengeType = challenge.name;
            const buttonColor =
              challengeType === selectedChallengeType ? "red.500" : "blue.500";
            console.log(selectedChallengeType, challengeType);
            return (
              <Button
                testID="option"
                backgroundColor={buttonColor}
                onPress={() => setSelectedChallengeType(challengeType)}
                key={challengeType}
              >
                {challengeType}
              </Button>
            );
          })}
      </HStack>
    </StatusContainer>
  );
};

export default LeaderboardSettings;
