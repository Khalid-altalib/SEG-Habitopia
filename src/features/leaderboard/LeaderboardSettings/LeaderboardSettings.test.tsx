import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react-native";
import LeaderboardSettings from "./LeaderboardSettings";
import configureStore from "redux-mock-store";
import TestingWrapperNavigation from "@app/testingWrapperWithNavigation";

const mockStore = configureStore([]);
jest.mock("@features/challenges/challengesSlice", () => ({
  fetchChallenges: jest.fn(),
}));
describe("LeaderboardSettings", () => {
  let store: any;
  let wrapper: any;

  beforeEach(() => {
    store = mockStore({
      leaderboard: {
        timeInterval: "Weekly",
      },
      challenges: {
        challenges: [
          { name: "Challenge 1" },
          { name: "Challenge 2" },
          { name: "Challenge 3" },
        ],
        fetchChallenges: {
          loading: false,
          error: null,
        },
      },
    });
    wrapper = (
      <TestingWrapperNavigation store={store}>
        <LeaderboardSettings />
      </TestingWrapperNavigation>
    );
  });

  it("renders leaderboardSettings", () => {
    const { getByTestId } = render(wrapper);
    expect(getByTestId("leaderboardSettings")).toBeDefined();
  });

  it("matches snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });
});
