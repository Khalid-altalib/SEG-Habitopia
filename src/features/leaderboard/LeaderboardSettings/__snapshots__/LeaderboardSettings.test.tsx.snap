// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LeaderboardSettings matches snapshot 1`] = `
<TestingWrapperNavigation
  store={
    {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <LeaderboardSettings />
</TestingWrapperNavigation>
`;
